core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -287,6 +289,36 @@ public static BigDecimal rescale(BigDecimal value, DecimalType type)
         return value;
     }
 
+    public static long roundDown(Long value, int fromScale, int toScale)
+    {
+        // Use MAX_SHORT_PRECISION to fit all precision long values
+        if (fromScale < toScale) {
+            throw new IllegalArgumentException("source scale should be larger than target scale for rounding down");
+        }
+        return shortToShortCast(
+                value,
+                MAX_SHORT_PRECISION,
+                fromScale,
+                MAX_SHORT_PRECISION,
+                toScale,
+                longTenToNth(fromScale - toScale),
+                longTenToNth(fromScale - toScale - 1) * 5);
```

Comment:
why not `/2`

Reply:
Since this is a performance-sensitive area, I tried to use multiplication instead of division. But yes, division by 2 will fit in better here both semantically and operationally.